crypto/s2n_dhe.c

```diff
@@ -134,18 +134,20 @@ int s2n_check_all_dh_params(struct s2n_dh_params *dh_params)
     return S2N_SUCCESS;
 }
 
+DEFINE_POINTER_CLEANUP_FUNC(DH *, DH_free);
 int s2n_pkcs3_to_dh_params(struct s2n_dh_params *dh_params, struct s2n_blob *pkcs3)
 {
     POSIX_ENSURE_REF(dh_params);
     POSIX_PRECONDITION(s2n_blob_validate(pkcs3));
 
     uint8_t *original_ptr = pkcs3->data;
-    dh_params->dh = d2i_DHparams(NULL, (const unsigned char **) (void *) &pkcs3->data, pkcs3->size);
+    DEFER_CLEANUP(DH *dh_val = d2i_DHparams(NULL, (const unsigned char **) (void *) &pkcs3->data, pkcs3->size), DH_free_pointer);
+    dh_params->dh = dh_val;
```

Comment:
I don't love this. On error you'll free dh_val but dh_params->dh will still point to the (now invalid) memory. HOWEVER, it looks like that was actually true before, since DH_free(dh_params->dh) can't change the value of dh_params->dh and we never set it to NULL. So I guess this works.

Alternatively, you might want to use a temporary `s2n_dh_params` and [s2n_dh_params_free](https://github.com/aws/s2n-tls/blob/023618cfa6214bde0d235ac9ef4b35e0489c8c7b/crypto/s2n_dhe.c#L336C1-L336C1) and only set the value of dh_params at the end of the method. But that relies on the fact that you know `s2n_dh_params` is just a wrapper around `DH*`. However, arguably this method already relies on that fact.

Minimally, you probably need a comment here.

Reply:
We talked offline, and due to the lack of testing here I think we have to stick to the safest memory management model we can. We we really need to use the alternative method I suggested, which is how we generally manage memory for constructor / setup methods.